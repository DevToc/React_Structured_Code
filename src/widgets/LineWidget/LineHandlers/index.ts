import Straight from './Straight';

import { LineWidgetTypes } from '../LineWidget.types';

export const LINE_HANDLER_MAP = {
  [LineWidgetTypes.straight]: Straight,
  [LineWidgetTypes.curved]: null,
  [LineWidgetTypes.stepped]: null,
};
