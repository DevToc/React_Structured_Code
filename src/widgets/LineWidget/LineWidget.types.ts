import { Widget, AccessibleElement } from '../../types/widget.types';
import { WidgetId } from '../../types/idTypes';

export enum LineWidgetTypes {
  straight = 'straight',
  stepped = 'stepped',
  curved = 'curved',
}

export enum LineWidgetDashArrayTypes {
  solid = 'solid',
  dotted = 'dotted',
  dashed = 'dashed',
  dashDot = 'dashDot',
  longDash = 'longDash',
  longDashDot = 'longDashDot',
}

export enum ArrowStyleTypes {
  none = 'none',
  basic = 'basic',
  long = 'long',
  sharp = 'sharp',
}

/**
 * PortPosition Component
 */
export enum Alignment {
  BEGIN = 'BEGIN',
  CENTER = 'CENTER',
  END = 'END',
}

export enum Side {
  CENTER = 'CENTER',
  EAST = 'EAST',
  WEST = 'WEST',
  SOUTH = 'SOUTH',
  NORTH = 'NORTH',
}

export type LineWidgetTypeKeys = keyof typeof LineWidgetTypes;
export type LineWidgetDashArrayTypeKeys = keyof typeof LineWidgetDashArrayTypes;
export type ArrowStyleTypeKeys = keyof typeof ArrowStyleTypes;

/**
 * Extra data which needs to be saved to backend
 */
export interface Pos {
  xPx: number;
  yPx: number;
}

export interface DeltaPos extends Pos {
  targetIndex?: number;
}

export interface AnglePos extends Pos {
  angle: number;
  distance: number;
}

// For connection lines
export interface PortData {
  widgetId: WidgetId;
  alignment: Alignment;
  side: Side;
  offsetPx?: number; // When using a detailed offset to the port location
  offsetDeg?: number; // for center
}

interface LineConnection {
  startPort: PortData | null;
  endPort: PortData | null;
}

export interface LineWidgetData extends Widget, AccessibleElement, LineConnection {
  type: LineWidgetTypes;
  startPos: Pos;
  midPosList: Pos[];
  endPos: Pos;
  strokeWidth: number;
  strokeColor: string;
  strokeDashType: LineWidgetDashArrayTypeKeys;
  startArrowStyle: ArrowStyleTypeKeys;
  endArrowStyle: ArrowStyleTypeKeys;
}

export interface StraightLineProps {
  widgetId?: WidgetId;
  widthPx?: number;
  heightPx?: number;
  posList: Pos[];
  strokeWidth: number;
  strokeColor: string;
  strokeDashType: LineWidgetDashArrayTypeKeys;
  startArrowStyle: ArrowStyleTypeKeys;
  endArrowStyle: ArrowStyleTypeKeys;
  delta?: DeltaPos;
  isReadonly?: boolean;
}

export interface DragHandlerProps extends LineConnection {
  widgetId?: string;
  leftPx: number;
  topPx: number;
  startPort: PortData | null;
  endPort: PortData | null;
  posList: Pos[];
  zoom: number;
  handlerSize: number;
  activeHandlerIndex: number;
  setFocus: Function;
  delta: DeltaPos;
  setDelta: Function;
  updateData: Function;
}

export enum WidgetState {
  Default = 'default',
  Active = 'active',
  Edit = 'edit',
}
export interface LineWidgetControlKeyboardShortcutsProps {
  setWidgetState: (arg: WidgetState) => void;
  widgetState: WidgetState;
  setFocus: Function;
  onKeyEvent: Function;
}
export interface SetFocusProps {
  moveNext?: boolean;
  movePrev?: boolean;
  index?: number;
  isWidget?: boolean;
}
