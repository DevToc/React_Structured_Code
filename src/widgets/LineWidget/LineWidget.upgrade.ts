// the latest version of the widget
export const VERSION = 1;

// Add VersionControllers here that implement the WidgetVersionController interface
export const lineWidgetVersionControllers = [];
