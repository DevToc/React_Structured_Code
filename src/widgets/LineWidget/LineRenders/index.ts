import Straight from './Straight';

import { LineWidgetTypes } from '../LineWidget.types';

export const LINE_TYPE_MAP = {
  [LineWidgetTypes.straight]: Straight,
  [LineWidgetTypes.curved]: null,
  [LineWidgetTypes.stepped]: null,
};
