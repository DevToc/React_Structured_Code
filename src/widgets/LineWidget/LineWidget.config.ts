import { DeltaPos, LineWidgetTypes, LineWidgetDashArrayTypes, ArrowStyleTypes } from './LineWidget.types';
import { Key } from '../../constants/keyboard';

export const BOUNDING_BOX_CONFIG = {
  hideDefaultLines: true,
  resizable: false,
  rotatable: false,
};

export const DEFAULT_DELTA_POS: DeltaPos = {
  xPx: 0,
  yPx: 0,
};

// For the line renderer
export const DEFAULT_TYPE = LineWidgetTypes.straight;
export const DEFAULT_WIDTH = 400;
export const DEFAULT_HEIGHT = 400;
export const DEFAULT_FILL_COLOR = '#000';
export const DEFAULT_STROKE_WIDTH = 3;
export const DEFAULT_STROKE_DASH_TYPE = LineWidgetDashArrayTypes.solid;
export const DEFAULT_STROKE_START_ARROW_STYLE = ArrowStyleTypes.none;
export const DEFAULT_STROKE_END_ARROW_STYLE = ArrowStyleTypes.none;
export const DEFAULT_SHIFT_SNAP_ANGLE = 45;
export const DEFAULT_STROKE_SELECTION_TARGET = 10;

// For the custom handler
export const DEFAULT_CUSTOM_HANDLE_SIZE = 10;

// Minimum width or height that a line widget can have
export const MIN_WIDGET_SIZE = 20;

export const HANDLER_PADDING = 0;

export const KEYBOARD_CONFIG = {
  excludeSpecific: [Key.Enter],
};
