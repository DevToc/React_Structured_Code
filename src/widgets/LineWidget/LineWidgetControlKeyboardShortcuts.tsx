import { Key } from '../../constants/keyboard';
import { useEventListener } from '../../hooks/useEventListener';
import { FocusArea } from '../../modules/Editor/store/editorSettingsSlice.types';
import { WidgetState } from './LineWidget.types';
import { LineWidgetControlKeyboardShortcutsProps } from './LineWidget.types';
import { useEditor } from 'widgets/sdk';

export const LineWidgetControlKeyboardShortcuts = ({
  setWidgetState,
  widgetState,
  setFocus,
  onKeyEvent,
}: LineWidgetControlKeyboardShortcutsProps) => {
  const { focusArea, isWidgetFocused } = useEditor();

  const isToolbarFocused = focusArea === FocusArea.Toolbar;
  const isEditable = widgetState === WidgetState.Edit;
  const isActive = widgetState === WidgetState.Active;
  const isDefault = widgetState === WidgetState.Default;

  const onKeyDown = (e: KeyboardEvent | Event) => {
    const event = e as KeyboardEvent;
    const key: Key = event.which || event.keyCode;
    const isMod: boolean = event.metaKey || event.ctrlKey;
    const isShift: boolean = event.shiftKey;

    if (isEditable && isMod) {
      setFocus({ isWidget: true });
      setWidgetState(WidgetState.Active);
      return;
    }

    switch (key) {
      case Key.Escape:
        if (isEditable) {
          setFocus({ isWidget: true });
          setWidgetState(WidgetState.Active);
        }
        break;
      case Key.Enter:
        if (isActive || isDefault) {
          if (isWidgetFocused()) {
            setWidgetState(WidgetState.Edit);
            setFocus({ index: 0 });
          }
        }
        break;
      case Key.Tab:
        if (isEditable && !isToolbarFocused) {
          e.preventDefault();
          if (isShift) return setFocus({ movePrev: true });
          return setFocus({ moveNext: true });
        }
        break;
      case Key.UpArrow:
      case Key.DownArrow:
      case Key.LeftArrow:
      case Key.RightArrow:
        if (isMod) return;
        if (isEditable && !isToolbarFocused) {
          e.preventDefault();
          onKeyEvent(key, isShift);
        }
        break;
    }
  };

  useEventListener('keydown', onKeyDown);

  return <></>;
};
