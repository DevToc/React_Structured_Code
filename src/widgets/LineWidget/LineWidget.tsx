import { ReactElement, useEffect, useState, useMemo, useCallback } from 'react';
import { Box } from '@chakra-ui/react';

import {
  CustomOnResize,
  CustomOnResizeEnd,
  CustomOnRotateEnd,
} from 'modules/Editor/components/BoundingBox/BoundingBox.types';
import {
  useEditor,
  useWidget,
  WidgetToolbar,
  WidgetBase,
  WidgetBaseProp,
  ReadOnlyWidgetBase,
  TransparentImg,
} from 'widgets/sdk';
import { Pos, DeltaPos, LineWidgetData, WidgetState, SetFocusProps } from './LineWidget.types';
import { DEFAULT_DELTA_POS } from './LineWidget.config';
import { LINE_TYPE_MAP } from './LineRenders';
import { LINE_HANDLER_MAP } from './LineHandlers';
import {
  mergePosList,
  getStartPos,
  getMidPosList,
  getEndPos,
  getStepByKeyEvent,
  getPosByKeyEvent,
  getHandleSize,
  calculatePosByRatio,
  calculateWidgetRect,
  calculatePosListByAngle,
  calculateTargetPoint,
} from './LineWidget.helpers';
import { LineWidgetToolbarMenu } from './LineWidgetToolbar';
import { LineWidgetControlKeyboardShortcuts } from './LineWidgetControlKeyboardShortcuts';
import { Key } from '../../constants/keyboard';
import { useFlowCoreValue } from 'modules/Editor/components/FlowCore';
import { usePrevious } from 'hooks/usePrevious';
import { OnChangeEventTypes } from 'types/flowCore.types';

const LineWidget = (): ReactElement => {
  const {
    leftPx,
    topPx,
    widthPx,
    heightPx,
    startPos,
    midPosList,
    endPos,
    type,
    strokeWidth,
    strokeColor,
    strokeDashType,
    startArrowStyle,
    endArrowStyle,
    rotateDeg,
    widgetId,
    startPort,
    endPort,
    updateWidget,
  } = useWidget<LineWidgetData>();

  const [delta, setDelta] = useState<DeltaPos>(DEFAULT_DELTA_POS);
  const posListProps = useMemo(() => mergePosList(startPos, midPosList, endPos), [startPos, midPosList, endPos]);
  const [posListState, setPosListState] = useState(posListProps);
  const prevPosListProps = usePrevious(posListState);
  const [activeHandlerIndex, setActiveHandlerIndex] = useState<number>(-1);
  const [widgetState, setWidgetState] = useState(WidgetState.Default);

  const {
    zoom,
    isWidgetSelected,
    isGroupMember,
    setWidgetFocus,
    blur,
    enableEditorKeyboardShortcuts,
    disableEditorKeyboardShortcuts,
  } = useEditor();

  const handlerSize = useMemo(() => getHandleSize(zoom), [zoom]);
  const Line = useMemo(() => LINE_TYPE_MAP[type] as React.ElementType, [type]);
  const LineHandler = useMemo(() => LINE_HANDLER_MAP[type] as React.ElementType, [type]);

  const isPosListStateUpdated = useMemo(() => prevPosListProps !== posListState, [prevPosListProps, posListState]);
  const posList = useMemo(() => {
    let result = isPosListStateUpdated ? posListState : posListProps;

    if (rotateDeg !== 0) {
      result = calculatePosListByAngle(widthPx, heightPx, result, rotateDeg);
    }

    return result;
  }, [rotateDeg, widthPx, heightPx, isPosListStateUpdated, posListProps, posListState]);

  const flowCoreEvent = useFlowCoreValue('flowCoreEvent');

  const calculateDelta = useCallback(
    (matrix: DOMMatrix, target: HTMLElement): DeltaPos => {
      const observingWidgetId = target.id;
      const isStartPort = startPort?.widgetId === observingWidgetId;
      const isEndPort = endPort?.widgetId === observingWidgetId;
      const targetPortData = isStartPort ? startPort : isEndPort ? endPort : null;

      const targetIndex = isStartPort ? 0 : isEndPort ? posList.length - 1 : null;
      if (targetIndex === null || targetPortData === null) return DEFAULT_DELTA_POS;

      const targetPoint = calculateTargetPoint(target, matrix, targetPortData);
      const linePoint = [leftPx + posList[targetIndex].xPx, topPx + posList[targetIndex].yPx];
      const customDelta = [targetPoint[0] - linePoint[0], targetPoint[1] - linePoint[1]];

      const result = {
        targetIndex,
        xPx: customDelta[0],
        yPx: customDelta[1],
      };
      return result;
    },
    [startPort, endPort, leftPx, topPx, posList],
  );

  // NOTE: Do not use 'delta'
  // Since the delta value is state so that it has the race condition issue with moveablajs events
  const generateLineWidgetData = useCallback(
    (matrix: DOMMatrix, target: HTMLElement) => {
      const calcDelta = calculateDelta(matrix, target);
      const result = calculateWidgetRect(leftPx, topPx, calcDelta, posList);
      setDelta(DEFAULT_DELTA_POS);
      return result;
    },
    [leftPx, topPx, posList, calculateDelta, setDelta],
  );

  // Subscribe this line widget events to the FlowCoreContext
  useEffect(() => {
    flowCoreEvent.onChange(widgetId, (type, matrix, target) => {
      if (type === OnChangeEventTypes.Update) {
        // Update: The event type that is called from the Mutation Callback when dragging the connected widget
        setDelta(calculateDelta(matrix, target));
      } else if (type === OnChangeEventTypes.Done) {
        // Done: A callback function called from onDragEnd or onResizeEnd, which calculates and returns the current line widget data.
        return generateLineWidgetData(matrix, target);
      }
    });
  }, [widgetId, flowCoreEvent, calculateDelta, generateLineWidgetData]);

  // Custom update data function for custom handlers of the line widget
  const handleUpdateData = useCallback(
    (data: LineWidgetData) => {
      updateWidget(data);
      setDelta(DEFAULT_DELTA_POS);
    },
    [updateWidget],
  );

  // Move custom handler through key event
  const handleKeyEvent = useCallback(
    (key: Key, isShift: boolean) => {
      const delta: DeltaPos = {
        targetIndex: activeHandlerIndex,
        ...getPosByKeyEvent(key, getStepByKeyEvent(isShift)),
      };
      const newWidgetData = calculateWidgetRect(leftPx, topPx, delta, posList);
      if (newWidgetData) {
        handleUpdateData(newWidgetData);
        setPosListState(mergePosList(newWidgetData.startPos, newWidgetData.midPosList, newWidgetData.endPos));
      }
    },
    [leftPx, topPx, posList, activeHandlerIndex, handleUpdateData],
  );

  // Custom onResize handler to update the arrow line positions to allow for live rendering
  const customOnResize = useCallback(
    ({ event, onResize }: CustomOnResize) => {
      // Ignore keyboard shortcut
      if (!event?.inputEvent) return;

      const ratio = event.width / widthPx;
      const scaledStartPos: Pos = calculatePosByRatio(startPos, ratio);
      const scaledMidPosList: Pos[] = midPosList.map((pos) => calculatePosByRatio(pos, ratio));
      const scaledEndPos: Pos = calculatePosByRatio(endPos, ratio);
      const scaledPosList: Pos[] = mergePosList(scaledStartPos, scaledMidPosList, scaledEndPos);
      const rotatedPosList = calculatePosListByAngle(widthPx, heightPx, scaledPosList, -rotateDeg);

      setPosListState(rotatedPosList);

      onResize(event);
    },
    [startPos, midPosList, endPos, widthPx, heightPx, rotateDeg],
  );

  // trigger onResizeEnd with resized pos data.
  const customOnResizeEnd = useCallback(
    ({ event, onResizeEnd }: CustomOnResizeEnd) => {
      // Ignore keyboard shortcut
      if (!event?.inputEvent) return;

      onResizeEnd(event, {
        startPos: getStartPos(posList),
        midPosList: getMidPosList(posList),
        endPos: getEndPos(posList),
      });
    },
    [posList],
  );

  /**
   * Do not trigger onRotateEnd to prevent to save from the bounding box event
   */
  const customOnRotateEnd = useCallback(
    ({ event, onRotateEnd, isGroup }: CustomOnRotateEnd) => {
      // For the group rotate event, use the rotate event in the bounding box
      if (isGroup) {
        onRotateEnd(event);
        return;
      }

      const angle = event.lastEvent.rotate;
      const newPosList = calculatePosListByAngle(widthPx, heightPx, posList, angle);
      const newWidgetData = calculateWidgetRect(leftPx, topPx, delta, newPosList);
      if (newWidgetData) {
        newWidgetData.rotateDeg = 0;
        handleUpdateData(newWidgetData);

        // Reset bounding box changes
        event.target.style.transform = '';
      }
    },
    [posList, delta, heightPx, widthPx, leftPx, topPx, handleUpdateData],
  );

  const handleFocus = useCallback(
    ({ moveNext, movePrev, index, isWidget }: SetFocusProps) => {
      // Reset the activeElement
      blur();

      // Change focus to the widget
      if (isWidget) {
        setWidgetFocus();
        setActiveHandlerIndex(-1);
        if (widgetState !== WidgetState.Active) setWidgetState(WidgetState.Active);
        return;
      }

      // Change focus to the custom handler
      if (moveNext) {
        const nextIndex = posList.length <= activeHandlerIndex + 1 ? 0 : activeHandlerIndex + 1;
        setActiveHandlerIndex(nextIndex);
      } else if (movePrev) {
        const prevIndex = activeHandlerIndex - 1 < 0 ? posList.length - 1 : activeHandlerIndex - 1;
        setActiveHandlerIndex(prevIndex);
      } else if (typeof index === 'number') {
        setActiveHandlerIndex(index);
      }
      setWidgetState(WidgetState.Edit);
    },
    [widgetState, activeHandlerIndex, posList, setWidgetFocus, blur],
  );

  /**
   * Change the widgetState and activeElement(focus) according to the isWidgetSelected
   * Default - Rendered widget without focusing
   * Active - Selected widget with focusing
   * Edit - Selected the custom handler with focusing
   */
  useEffect(() => {
    // Check whether to use key event according to the widgetState
    if (widgetState === WidgetState.Default || widgetState === WidgetState.Active) enableEditorKeyboardShortcuts();
    if (isWidgetSelected && widgetState === WidgetState.Edit) disableEditorKeyboardShortcuts();

    // Change the widgetState and initialize the handler index according to the current state
    if (!isWidgetSelected && widgetState !== WidgetState.Default) setWidgetState(WidgetState.Default);
    if (isWidgetSelected && widgetState === WidgetState.Default) {
      setWidgetState(WidgetState.Active);
      setActiveHandlerIndex(-1);
    }
  }, [isWidgetSelected, widgetState, enableEditorKeyboardShortcuts, disableEditorKeyboardShortcuts]);

  return (
    <WidgetBase
      onResize={customOnResize}
      onResizeEnd={customOnResizeEnd}
      onRotateEnd={customOnRotateEnd}
      disableSingleSelect
    >
      {/* pointerEvents is disabled to use the custom selection area in the line renderer */}
      <Box
        w='100%'
        h='100%'
        position='relative'
        data-testid='line-widget'
        pointerEvents='none'
        style={{ transform: isPosListStateUpdated ? 'rotate(0deg)' : `rotate(${-rotateDeg}deg)` }}
      >
        <WidgetToolbar>
          <LineWidgetToolbarMenu />
        </WidgetToolbar>
        {isWidgetSelected && (
          <LineWidgetControlKeyboardShortcuts
            setFocus={handleFocus}
            setWidgetState={setWidgetState}
            widgetState={widgetState}
            onKeyEvent={handleKeyEvent}
          />
        )}
        <Line
          posList={posList}
          widgetId={widgetId}
          widthPx={widthPx}
          heightPx={heightPx}
          delta={delta}
          strokeWidth={strokeWidth}
          strokeColor={strokeColor}
          strokeDashType={strokeDashType}
          startArrowStyle={startArrowStyle}
          endArrowStyle={endArrowStyle}
        />

        {/* Drag handler */}
        {isWidgetSelected && !isGroupMember && (
          <LineHandler
            widgetId={widgetId}
            posList={posList}
            leftPx={leftPx}
            topPx={topPx}
            delta={delta}
            zoom={zoom}
            handlerSize={handlerSize}
            activeHandlerIndex={activeHandlerIndex}
            startPort={startPort}
            endPort={endPort}
            setFocus={handleFocus}
            setDelta={setDelta}
            updateData={handleUpdateData}
          />
        )}
      </Box>
    </WidgetBase>
  );
};

const ReadOnlyLineWidget = ({ includeAltTextImg }: WidgetBaseProp): ReactElement => {
  const {
    widthPx,
    heightPx,
    startPos,
    midPosList,
    endPos,
    type,
    strokeWidth,
    strokeColor,
    strokeDashType,
    startArrowStyle,
    endArrowStyle,
    widgetId,
  } = useWidget<LineWidgetData>();

  const posList = useMemo(() => mergePosList(startPos, midPosList, endPos), [startPos, midPosList, endPos]);
  const Line = LINE_TYPE_MAP[type] as React.ElementType;

  return (
    <ReadOnlyWidgetBase>
      <Box w='100%' h='100%' position='relative' data-testid='line-widget'>
        {includeAltTextImg && <TransparentImg />}
        <Line
          posList={posList}
          widgetId={widgetId}
          widthPx={widthPx}
          heightPx={heightPx}
          delta={DEFAULT_DELTA_POS}
          strokeWidth={strokeWidth}
          strokeColor={strokeColor}
          strokeDashType={strokeDashType}
          startArrowStyle={startArrowStyle}
          endArrowStyle={endArrowStyle}
          isReadonly
        />
      </Box>
    </ReadOnlyWidgetBase>
  );
};

export { LineWidget, ReadOnlyLineWidget };
