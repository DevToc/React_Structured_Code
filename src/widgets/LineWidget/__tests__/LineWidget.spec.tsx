import { screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { LineWidget } from '../LineWidget';
import { renderWithRedux, mockAddLineWidget, mockSetActiveWidget, renderWidget } from 'utils/test-utils.test';
import { generateDefaultData } from '../LineWidget.helpers';
import { LineWidgetTypes } from '../LineWidget.types';
import { SELECTO_TARGET_CLASS } from '../../../constants/bounding-box';
import { EMPTY_INFOGRAPH } from '../../../utils/loadSampleInfograph';
import { InfographLoader } from '../../../modules/InfographLoader';
import Editor from '../../../modules/Editor';

describe('widgets/LineWidget', () => {
  it('should render correctly', () => {
    const widgetId = '005.widget-1';
    const widget = { widgetId, ...generateDefaultData(LineWidgetTypes.straight) };
    const { asFragment } = renderWidget(<LineWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render line widget', () => {
    const widgetId = '005.widget-1';
    const defaultData = generateDefaultData(LineWidgetTypes.straight);

    const widget = {
      widgetId,
      ...defaultData,
      widgetData: {
        ...defaultData.widgetData,
        altText: null,
        isDecorative: true,
      },
    };

    renderWidget(<LineWidget />, { widget });

    expect(screen.getByTestId('line-widget')).toBeInTheDocument();
    expect(screen.getByTestId('line-widget-renderer')).toBeInTheDocument();

    // To use custom selection, there must be no single selection target class.
    const widgetElement = screen.getByTestId(`widgetbase-${widgetId}`);
    expect(widgetElement.classList.contains(SELECTO_TARGET_CLASS)).toBeTruthy();
    expect(widgetElement).toHaveStyle({ pointerEvents: 'none' });
  });

  it('should provide the a11y mode via key input', () => {
    renderWithRedux(
      <InfographLoader infographState={EMPTY_INFOGRAPH}>
        <Editor />
      </InfographLoader>,
    );

    mockAddLineWidget();
    const widget = screen.queryAllByTestId(/widgetbase-/)[0];
    mockSetActiveWidget(widget.id);

    const [startHandler, endHandler] = screen.getAllByTestId('linewidget-customhandler');
    expect(startHandler).toHaveAttribute('data-handler-direction', 'start');
    expect(startHandler).toHaveAttribute('data-handler-index', '0');
    expect(endHandler).toHaveAttribute('data-handler-direction', 'end');
    expect(endHandler).toHaveAttribute('data-handler-index', '1');

    // It use fireEvent instead of the userEvent because of the pointer-events none on the widget base
    fireEvent.click(widget);

    // When entering edit mode for the first time, focusing with start handler
    userEvent.keyboard('{enter}');
    expect(startHandler).toHaveFocus();

    // When triggering tab key, it moves the focus to the next handler
    userEvent.tab();
    expect(endHandler).toHaveFocus();

    // When triggering tab key and it is the last handler, it moves the focus to the first handler
    userEvent.tab();
    expect(startHandler).toHaveFocus();

    // Reset the focus to the widget when triggering the escape key
    userEvent.keyboard('{esc}');
    expect(widget).toHaveFocus();
  });
});
