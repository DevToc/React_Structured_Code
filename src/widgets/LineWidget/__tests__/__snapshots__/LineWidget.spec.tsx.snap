// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/LineWidget should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="005.widget-1-description"
    aria-label="Line widget"
    class="widget-container widget-selecto-target  css-1pzw6kx"
    data-testid="widgetbase-005.widget-1"
    id="005.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 400px; height: 400px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-1a3qu3o"
      data-testid="line-widget"
      style="transform: rotate(0deg);"
    >
      <svg
        data-testid="line-widget-renderer"
        style="width: 400px; height: 400px; overflow: visible; position: absolute;"
      >
        <defs>
          <marker
            id="start-arrow-005.widget-1"
            orient="auto-start-reverse"
            refX="10px"
            refY="5px"
            viewBox="0 0 10 10"
          >
            <path
              d=""
              fill="#000"
            />
          </marker>
          <marker
            id="end-arrow-005.widget-1"
            orient="auto"
            refX="10px"
            refY="5px"
            viewBox="0 0 10 10"
          >
            <path
              d=""
              fill="#000"
            />
          </marker>
        </defs>
        <path
          d="M 0 0 L 400 400"
          fill="transparent"
          stroke="#000"
          stroke-dasharray="none"
          stroke-width="3px"
        />
        <path
          d="M 0 0 L 400 400"
          fill="transparent"
          marker-end="url(#end-arrow-005.widget-1)"
          marker-start="url(#start-arrow-005.widget-1)"
          stroke="transparent"
          stroke-width="3px"
        />
        <path
          d="M 0 0 L 400 400"
          fill="transparent"
          id="005.widget-1"
          pointer-events="stroke"
          stroke="transparent"
          stroke-dasharray="none"
          stroke-width="10px"
        />
      </svg>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="005.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
